merkle/hstar2.go

```diff
@@ -83,19 +87,51 @@ func (s *HStar2) HStar2Nodes(prefix []byte, subtreeDepth int, values []*HStar2Le
 			glog.Infof("  %x: %x", v.Index.Bytes(), v.LeafHash)
 		}
 	}
+	combine := func(depth int, offset *big.Int, lhs, rhs []byte) ([]byte, error) {
+		h := s.hasher.HashChildren(lhs, rhs)
+		if err := s.set(offset, depth, h, set); err != nil {
+			return nil, err
+		}
+		return h, nil
+	}
+	return s.run(prefix, subtreeDepth, values, get, combine)
+}
+
+// Prefetch does a dry run of HStar2 algorithm, and reports all Merkle tree
+// nodes that it needs through the passed-in visit function. Note that the
+// return value of the visit function is ignored, unless it is an error.
+//
+// This function can be useful, for example, if the caller prefers to collect
+// the node IDs and read them from storage in one batch. Then they can run
+// HStar2Nodes in such a way that it reads from the prefetched set.
+func (s *HStar2) Prefetch(prefix []byte, subtreeDepth int, values []*HStar2LeafHash, visit SparseGetNodeFunc) error {
+	combine := func(depth int, offset *big.Int, lhs, rhs []byte) ([]byte, error) {
+		return nil, nil
+	}
+	_, err := s.run(prefix, subtreeDepth, values, visit, combine)
+	return err
+}
+
+// combineFunc returns a node hash based on two child hashes. It may also do
+// side effects, e.g. put the resulting node to storage.
+type combineFunc func(depth int, offset *big.Int, lhs, rhs []byte) ([]byte, error)
```

Comment:
It took me a little while to work out why this function was needed over the logically cleaner "set" function. My calculated answer is that in the prefetch case you don't want to actually be doing any lookups for node values in-line, and thus you don't have any hashes to combine. By passing in the combine function you can avoid the nil hash error. Consider making this explicit in this comment, e.g. "this method may perform no work if none is needed, such as the prefetch case."

Reply:
Done.